units/en/unit2/hands-on.mdx

```diff
@@ -0,0 +1,1074 @@
+- Create a new token (https://huggingface.co/settings/tokens) **with write role**
+
+
+<img src="https://huggingface.co/datasets/huggingface-deep-rl-course/course-images/resolve/main/en/notebooks/create-token.jpg" alt="Create HF Token">
+
+
+```python
+from huggingface_hub import notebook_login
+
+notebook_login()
+```
+
+If you don't want to use a Google Colab or a Jupyter Notebook, you need to use this command instead: `huggingface-cli login`
+
+3️⃣ We're now ready to push our trained agent to the 🤗 Hub 🔥 using `package_to_hub()` function
+
+- Let's create **the model dictionary that contains the hyperparameters and the Q_table**.
+
+```python
+model = {
+    "env_id": env_id,
+    "max_steps": max_steps,
+    "n_training_episodes": n_training_episodes,
+    "n_eval_episodes": n_eval_episodes,
+    "eval_seed": eval_seed,
+    "learning_rate": learning_rate,
+    "gamma": gamma,
+    "max_epsilon": max_epsilon,
+    "min_epsilon": min_epsilon,
+    "decay_rate": decay_rate,
+    "qtable": Qtable_frozenlake,
+}
+```
+
+Let's fill the `package_to_hub` function:
+
+- `repo_id`: the name of the Hugging Face Hub Repository that will be created/updated `
+(repo_id = {username}/{repo_name})`
+💡 A good `repo_id` is `{username}/q-{env_id}`
+- `model`: our model dictionary containing the hyperparameters and the Qtable.
+- `env`: the environment.
+- `commit_message`: message of the commit
+
+```python
+model
+```
+
+```python
+username = ""  # FILL THIS
+repo_name = "q-FrozenLake-v1-4x4-noSlippery"
+push_to_hub(repo_id=f"username}/{repo_name}", model=model, env=env)
+```
+
+Congrats 🥳 you've just implemented from scratch, trained and uploaded your first Reinforcement Learning agent.
+FrozenLake-v1 no_slippery is very simple environment, let's try an harder one 🔥.
+
+# Part 2: Taxi-v3 🚖
+
+## Create and understand [Taxi-v3 🚕](https://www.gymlibrary.dev/environments/toy_text/taxi/)
+---
+
+💡 A good habit when you start to use an environment is to check its documentation
+
+👉 https://www.gymlibrary.dev/environments/toy_text/taxi/
+
+---
+
+In `Taxi-v3` 🚕, there are four designated locations in the grid world indicated by R(ed), G(reen), Y(ellow), and B(lue).
+
+When the episode starts, **the taxi starts off at a random square** and the passenger is at a random location. The taxi drives to the passenger’s location, **picks up the passenger**, drives to the passenger’s destination (another one of the four specified locations), and then **drops off the passenger**. Once the passenger is dropped off, the episode ends.
+
+
+<img src="https://huggingface.co/datasets/huggingface-deep-rl-course/course-images/resolve/main/en/notebooks/unit2/taxi.png" alt="Taxi">
+
+
+```python
+env = gym.make("Taxi-v3")
+```
+
+There are **500 discrete states since there are 25 taxi positions, 5 possible locations of the passenger** (including the case when the passenger is in the taxi), and **4 destination locations.**
```

Comment:
```suggestion
There are **500 discrete states since there are 25 taxi positions and 5 possible locations of the passenger** (including the case when the passenger is in the taxi), and **4 destination locations.**
```